### PR TITLE
Fix a cuda max_pool3d issue, do multiplication in int64

### LINKED ISSUE #52822: Illegal memory access in cuda max pooling for large inputs
## 🐛 Bug

From #52211

```
import torch

x = torch.randn(70, 32, 100, 100, 100, dtype=torch.half, device='cuda')

y = torch.nn.functional.max_pool3d(x, 5)

torch.cuda.synchronize()
```
Env:

Collecting environment information...
PyTorch version: 1.9.0a0+a86027d
Is debug build: False
CUDA used to build PyTorch: 11.2
ROCM used to build PyTorch: N/A

OS: Manjaro Linux (x86_64)
GCC version: (GCC) 10.2.0
Clang version: Could not collect
CMake version: version 3.19.3

Python version: 3.9 (64-bit runtime)
Is CUDA available: True
CUDA runtime version: 11.2.67
GPU models and configuration: 
GPU 0: GeForce RTX 2070 SUPER
GPU 1: GeForce GTX 1070 Ti

Nvidia driver version: 460.32.03
cuDNN version: Probably one of the following:
/usr/lib/libcudnn.so.8.0.5
/usr/lib/libcudnn_adv_infer.so.8.0.5
/usr/lib/libcudnn_adv_train.so.8.0.5
/usr/lib/libcudnn_cnn_infer.so.8.0.5
/usr/lib/libcudnn_cnn_train.so.8.0.5
/usr/lib/libcudnn_ops_infer.so.8.0.5
/usr/lib/libcudnn_ops_train.so.8.0.5
HIP runtime version: N/A
MIOpen runtime version: N/A

Versions of relevant libraries:
[pip3] numpy==1.19.5
[pip3] torch==1.9.0a0+a86027d

Root cause https://github.com/pytorch/pytorch/issues/52211#issuecomment-779613083
cc @ezyang @gchanan @zou3519 @bdhirsh @jbschlosser @anjali411 @ngimel @heitorschueroff @xwang233

### PR DESCRIPTION
**Pull request by xwang233**
Fix #52822

- [x] benchmark

### CODE REVIEW

**Comment by xwang233** [comment]
Benchmark (basically no difference)

```
[------------------------- max_pool3d, kernel_size=2 -------------------------]
                                         |  1.9.0a0+0567988  |  1.9.0a0+c600c4b
1 threads: --------------------------------------------------------------------
      torch.Size([11, 22, 25, 8, 85])    |       173.8       |       171.5     
      torch.Size([7, 7, 110, 47, 24])    |       147.8       |       147.4     
      torch.Size([10, 7, 228, 21, 10])   |       194.0       |       193.9     
      torch.Size([8, 5, 29, 119, 64])    |       380.4       |       375.1     
      torch.Size([8, 11, 35, 12, 152])   |       199.3       |       200.4     
      torch.Size([6, 22, 121, 15, 33])   |       417.9       |       419.0     
      torch.Size([11, 11, 14, 43, 25])   |        84.0       |        84.2     
      torch.Size([16, 16, 18, 61, 23])   |       162.3       |       163.3     
      torch.Size([16, 38, 10, 127, 12])  |       563.5       |       566.6     
      torch.Size([6, 5, 110, 108, 28])   |       210.4       |       211.5     
      torch.Size([6, 9, 25, 17, 99])     |       111.5       |       112.4     
      torch.Size([4, 42, 28, 72, 22])    |       200.2       |       201.8     
      torch.Size([8, 54, 22, 15, 45])    |       324.9       |       324.6     
      torch.Size([4, 4, 211, 13, 48])    |       108.9       |       108.5     
      torch.Size([5, 127, 26, 8, 47])    |        99.7       |       100.4     
      torch.Size([6, 31, 79, 21, 31])    |       375.9       |       374.3     
      torch.Size([13, 8, 23, 24, 167])   |       125.3       |       125.8     
      torch.Size([5, 8, 26, 177, 25])    |       209.0       |       210.3     
      torch.Size([5, 132, 18, 62, 9])    |       377.3       |       380.6     
      torch.Size([4, 20, 23, 16, 256])   |       262.3       |       260.7     

Times are in microseconds (us).

[------------------------- max_pool3d, kernel_size=3 -------------------------]
                                         |  1.9.0a0+0567988  |  1.9.0a0+c600c4b
1 threads: --------------------------------------------------------------------
      torch.Size([11, 22, 25, 8, 85])    |       137.2       |       137.5     
      torch.Size([7, 7, 110, 47, 24])    |       120.7       |       121.1     
      torch.Size([10, 7, 228, 21, 10])   |       146.9       |       147.6     
      torch.Size([8, 5, 29, 119, 64])    |       355.8       |       356.7     
      torch.Size([8, 11, 35, 12, 152])   |       173.5       |       173.8     
      torch.Size([6, 22, 121, 15, 33])   |       422.5       |       422.9     
      torch.Size([11, 11, 14, 43, 25])   |        73.1       |        73.1     
      torch.Size([16, 16, 18, 61, 23])   |       141.9       |       142.5     
      torch.Size([16, 38, 10, 127, 12])  |       477.1       |       478.4     
      torch.Size([6, 5, 110, 108, 28])   |       176.7       |       177.3     
      torch.Size([6, 9, 25, 17, 99])     |       110.2       |       111.2     
      torch.Size([4, 42, 28, 72, 22])    |       152.2       |       152.6     
      torch.Size([8, 54, 22, 15, 45])    |       322.5       |       322.6     
      torch.Size([4, 4, 211, 13, 48])    |       109.3       |       109.1     
      torch.Size([5, 127, 26, 8, 47])    |        80.5       |        80.6     
      torch.Size([6, 31, 79, 21, 31])    |       337.1       |       336.1     
      torch.Size([13, 8, 23, 24, 167])   |        72.3       |        72.9     
      torch.Size([5, 8, 26, 177, 25])    |       192.4       |       193.2     
      torch.Size([5, 132, 18, 62, 9])    |       302.2       |       303.6     
      torch.Size([4, 20, 23, 16, 256])   |       242.9       |       241.4     

Times are in microseconds (us).

[------------------------- max_pool3d, kernel_size=5 -------------------------]
                                         |  1.9.0a0+0567988  |  1.9.0a0+c600c4b
1 threads: --------------------------------------------------------------------
      torch.Size([11, 22, 25, 8, 85])    |       159.9       |       160.0     
      torch.Size([7, 7, 110, 47, 24])    |       176.3       |       176.4     
      torch.Size([10, 7, 228, 21, 10])   |       157.5       |       157.6     
      torch.Size([8, 5, 29, 119, 64])    |       353.5       |       353.9     
      torch.Size([8, 11, 35, 12, 152])   |       167.3       |       167.5     
      torch.Size([6, 22, 121, 15, 33])   |       444.3       |       444.6     
      torch.Size([11, 11, 14, 43, 25])   |        68.6       |        68.9     
      torch.Size([16, 16, 18, 61, 23])   |       137.8       |       137.9     
      torch.Size([16, 38, 10, 127, 12])  |       552.8       |       553.5     
      torch.Size([6, 5, 110, 108, 28])   |       206.1       |       206.3     
      torch.Size([6, 9, 25, 17, 99])     |       112.7       |       113.5     
      torch.Size([4, 42, 28, 72, 22])    |       170.3       |       170.7     
      torch.Size([8, 54, 22, 15, 45])    |       342.8       |       342.9     
      torch.Size([4, 4, 211, 13, 48])    |       120.3       |       120.0     
      torch.Size([5, 127, 26, 8, 47])    |       109.1       |       109.0     
      torch.Size([6, 31, 79, 21, 31])    |       348.6       |       347.3     
      torch.Size([13, 8, 23, 24, 167])   |        68.8       |        68.6     
      torch.Size([5, 8, 26, 177, 25])    |       211.1       |       211.8     
      torch.Size([5, 132, 18, 62, 9])    |       274.6       |       275.0     
      torch.Size([4, 20, 23, 16, 256])   |       237.5       |       235.2     

Times are in microseconds (us).
```

<details>

<summary> code </summary>

<p>

b.py

```python
import torch
import torch.utils.benchmark as benchmark
import pickle

fuzzer = benchmark.Fuzzer(
    parameters=[
        benchmark.FuzzedParameter('n', minval=4, maxval=16, distribution='uniform'),
        benchmark.FuzzedParameter('c', minval=4, maxval=256, distribution='uniform'),
        benchmark.FuzzedParameter('d', minval=8, maxval=256, distribution='uniform'),
        benchmark.FuzzedParameter('h', minval=8, maxval=256, distribution='uniform'),
        benchmark.FuzzedParameter('w', minval=8, maxval=256, distribution='uniform'),
    ],
    tensors=[
        benchmark.FuzzedTensor(
            'x',
            size='ncdhw',
            min_elements=12,
            max_elements=10000000,
            cuda=True,
            dtype=torch.half,
            max_allocation_bytes=1_000_000_000
        )
    ],
    seed=42
)

res = []

for kernel_size in [2, 3, 5]:
    for tensors, tensor_params, params in fuzzer.take(20):
        sub_label = str(tensors['x'].size())
        res.append(
            benchmark.Timer(
                stmt=f'torch.nn.functional.max_pool3d(x, {kernel_size})',
                setup='',
                globals=tensors,
                label=f'max_pool3d, {kernel_size=}',
                sub_label=sub_label,
                description=f'{torch.__version__}'
            ).blocked_autorange(min_run_time=0.1)
        )

torch_ver = str(torch.__version__)
torch_git_ver = torch_ver[torch_ver.index('+') + 1:]

with open(f'{torch_git_ver}.pkl', 'wb') as f:
    pickle.dump(res, f)

compare = benchmark.Compare(res)
# compare.colorize()
compare.print()
```

c.py

```python
import torch.utils.benchmark as benchmark
import glob
import pickle

res = []

for pkl in glob.glob('./*.pkl'):
    with open(pkl, 'rb') as f:
        res += pickle.load(f)

compare = benchmark.Compare(res)
compare.print()

```


</p>

</details>

**Comment by xwang233** [comment]
cc @ptrblck

File: aten/src/ATen/native/cuda/DilatedMaxPool3d.cu [hunk 1]
```diff
@@ -53,7 +53,7 @@ __global__ static void max_pool3d_with_indices_single_out_frame(
       wStart += dilationW;
 
     int maxIndex =  tStart * iheight * iwidth + hStart * iwidth + wStart;
-    inputData += slice * itime * iheight * iwidth;
+    inputData += static_cast<int64_t>(slice) * itime * iheight * iwidth;
```
Comment (heitorschueroff) [inline aten/src/ATen/native/cuda/DilatedMaxPool3d.cu:56]: Maybe make slice itself int64_t instead of casting?

**Review by malfet — CHANGES_REQUESTED** [review]
Please rebase PR against `viable/strict` to get clean signal.
Also, can you please share end-to-end test run time somewhere? (Something tells me that `torch.randn(70, 32, 100, 100, 100)` will take 2+ min when run in ASAN)

**Comment by ngimel** [comment]
Do we have asan+cuda tests?

**Comment by xwang233** [comment]
If you mean the newly added test, it took 6s on my machine. Without CPU comparison and CUDA only, it took 1s.

**Comment by ngimel** [comment]
Ok, ci is clean, large tensor test won't be run in CI anyway, @heitorschueroff can you please land this PR?

**Pull request closed without merging**